src/stan/callbacks/file_stream_writer.hpp

```diff
@@ -0,0 +1,117 @@
+#ifndef STAN_CALLBACKS_FILE_STREAM_WRITER_HPP
+#define STAN_CALLBACKS_FILE_STREAM_WRITER_HPP
+
+#include <stan/callbacks/writer.hpp>
+#include <ostream>
+#include <vector>
+#include <string>
+
+namespace stan {
+namespace callbacks {
+
+/**
+ * <code>file_stream_writer</code> is an implementation
+ * of <code>writer</code> that writes to a file.
+ */
+class file_stream_writer final : public writer {
+ public:
+  /**
+   * Constructs a file stream writer with an output stream
+   * and an optional prefix for comments.
+   *
+   * @param[in, out] A unique pointer to a type inheriting from `std::ostream`
+   * @param[in] comment_prefix string to stream before each comment line.
+   *  Default is "".
+   */
+  explicit file_stream_writer(std::unique_ptr<std::ostream>&& output,
+                              const std::string& comment_prefix = "")
+      : output_(std::move(output)), comment_prefix_(comment_prefix) {}
+
+  file_stream_writer();
+  file_stream_writer(file_stream_writer& other) = delete;
```

Comment:
Oh okay I looked back at what you previously wrote:

```
So that file_stream_writer holds unique pointers to each of the file streams. I had some
issues using stream_writer where somewhere a reference was being dereferenced or
something was falling out of scope (stream_writer just holds it's ostream by refrerence)
so I thought the easiest solution was just to wrap that ostream into an std::unique_ptr.
Then we also get promises that no-one ever copies that thing.
```

It seems cleaner to replace stream_writer with something that behaves better. Presumably the other use cases of stream_writer can still work.

Reply:
Oh I see your previous comment on this:

```
So that file_stream_writer holds unique pointers to each of the file streams. I had some
issues using stream_writer where somewhere a reference was being dereferenced or
something was falling out of scope (stream_writer just holds it's ostream by refrerence)
so I thought the easiest solution was just to wrap that ostream into an std::unique_ptr.
Then we also get promises that no-one ever copies that thing.
```

Can we clean up the stream_writer implementation though instead of adding another thing that does something very similar?

Reply:
I'd like to replace stream writer, but I was having trouble with backwards compatibility. I think once cmdstan uses the `file_stream_writer` we can remove `stream_writer`

Reply:
Yeah that's the main issue